src/action/timeline.rs

```diff
@@ -0,0 +1,567 @@
+        let mut dir_iter_stack = Vec::new();
+        loop {
+            let metadata = match symlink_metadata(&path) {
+                Ok(metadata) => metadata,
+                Err(_) => continue,
+            };
+            let entry = entry_from_metadata(&metadata, &path).map_err(Error::action)?;
+            self.process_entry(entry, session)?;
+            if metadata.is_dir() && dev_from_metadata(&metadata) == self.device {
+                if let Ok(dir_iter) = read_dir(&path) {
+                    dir_iter_stack.push(dir_iter);
+                }
+            }
+            let mut new_path = None;
+            while let Some(dir_iter) = dir_iter_stack.last_mut() {
+                if let Some(dir_entry) = dir_iter.next() {
+                    new_path = Some(dir_entry.map_err(Error::action)?.path());
+                    break;
+                } else {
+                    dir_iter_stack.pop();
+                }
+            }
+            if let Some(new_path) = new_path {
+                path = new_path;
+            } else {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    /// Sends final pieces of data to the session.
+    fn finish<S: Session>(mut self, session: &mut S) -> session::Result<Vec<ChunkDigest>> {
+        let final_block = self.encoder.next_chunk().map_err(Error::action)?;
+        self.send_block(final_block, session)?;
+        Ok(self.ids)
+    }
+}
+
+/// Handles requests for the timeline action.
+pub fn handle<S: Session>(session: &mut S, request: Request) -> session::Result<()> {
+    let root_metadata = symlink_metadata(&request.root).map_err(Error::action)?;
+    let target_device = dev_from_metadata(&root_metadata);
+    let mut state = RecurseState::new(target_device);
+
+    state.recurse(&request.root, session)?;
+    let action_response = Response {
+        ids: state.finish(session)?,
+    };
+    session.reply(action_response)?;
+
+    Ok(())
+}
+
+impl super::Request for Request {
+
+    type Proto = TimelineArgs;
+
+    fn from_proto(proto: TimelineArgs) -> Result<Request, ParseError> {
+        match proto.root {
+            Some(root) => Ok(Request {
+                root: PathBuf::from(os_string_from_bytes(root.as_slice())),
+            }),
+            None => Err(ParseError::malformed(MissingFieldError::new("root"))),
+        }
+    }
+}
+
+impl super::Response for Response {
+
+    const RDF_NAME: Option<&'static str> = Some("TimelineResult");
+
+    type Proto = TimelineResult;
+
+    fn into_proto(self) -> TimelineResult {
+        TimelineResult {
+            entry_batch_blob_ids: self.ids.iter().map(|id| id.0.to_vec()).collect()
+        }
+    }
+}
+
+impl super::Response for ChunkResponse {
+
+    const RDF_NAME: Option<&'static str> = Some("DataBlob");
+
+    type Proto = DataBlob;
+
+    fn into_proto(self) -> DataBlob {
+        DataBlob {
+            data: Some(self.data),
+            ..Default::default()
+        }
```

Comment:
You should be able to use `self.data.into()` now.

Reply:
You mean `Vec` into `Some(Vec)`?
Why?

Reply:
I mean that now there is a `From<Vec<u8>>` [implemented](https://github.com/google/rrg/blob/cdec0598987acf03df22552223a5e91a0d25c370/proto/src/lib.rs#L38-L46) for `DataBlob`, so you can convert your `Vec<u8>` into `DataBlob` with just a single `.into()` call.